### ISSUE TITLE
Can't read webm's produced by MediaRecorder API with Chrome

### ISSUE THREAD

**Issue by Innominus** [austinleroy/webm-iterable#5]
Hey there,

I thought this might be the appropriate place to ask about this; love this library and the ebml-iterable library btw.
I'm currently trying to send video chunks from the MediaRecorder API to a backend with websockets, and then create individual playable webm files.

When iterating over webm's produced by FireFox I get correct output, Simpleblocks, cluster tags and segment tags.
When doing the same thing with Chrome's MediaRecorder API with the same mime type using VP9 encoding I get panics. Namely this panic on the initial chunk with the header file:
`thread 'main' panicked at 'slice index starts at 1 but ends at 0', C:\Users\kadth\.cargo\registry\src\github.com-1ecc6299db9ec823\ebml-iterable-0.4.0\src\tag_iterator.rs:131:29`
This panic always happens right after the Tracks(End) tag. 

If I try inspect the other chunks I just get rawtags instead of SimpleBlocks:
`RawTag(31493, [112, 159, 107, 219, 247, 162, 178, 240, 143, 179, 96, 251, 255, 23, 209, 237, 126, 91, 90, 229, 114, 101, 17, 126, 218, 254, 34, 131, 253, 151, 154, 150, 151, 146, 151, 87, 38, 215, 226, 122, 86, 185, 96, 163, 65, 122, 130, 6, 133, 0, 0, 0, 0, 1, 9, 48, 0, 0, 1, 78, 128, 128, 15, 32, 0, 0, 1, 65, 154, 100, 60, 2, 89, 204, 43, 146, 152, 153, 177, 175, 123, 12, 120, 37, 239, 71, 130, 126, 94, 83, 25, 55, 4, 124, 87, 123, 63, 238, 96, 128, 121, 249, 40, 248, 41, 229, 216, 120, 18, 164, 201, 46, 8, 45, 4, 110, 238])
Error: TagIterator(CorruptedFileData("reached end of file but expecting more data"))`
And they always seem to end with a CorruptedFileData error.

This is really strange to me. The header I manage to get before it errors out on the first file shows it's a webm container, and it's VP9 encoded, but this seems to not be the case. Are Google not following the spec correctly? 
Hoping someone would be able to shed some light on this. Thanks! :)

**Comment by austinleroy** [comment]
Hello! Thanks for your interest - I'm glad you've enjoyed using these libraries. I'd be happy to help try and figure out what's going on with the issue you mentioned. Do you happen to have an example binary I could debug with?

**Comment by Innominus** [comment]
@austinleroy 
Hey! Thanks so much for getting back to me. I've added you to my webm-restorer project. There are two folders, input and FF_input. The first input is input directly from the Chrome MediaRecorder API, and I've added FF_input from Firefox to show that FF is actually giving correct data. Should be able to just cargo run.
I've seen some posts out there explaining that the Chrome MediaRecorder for webm/VP9 is doing weird stuff such as putting the cluster tag at the end of a line, and maybe it's just out of order? Unsure. Let me know what you think.

**Comment by austinleroy** [comment]
So I had time to do some digging into this today, and I think I found something useful.  At least I hope it's useful and it isn't something you already knew 😅 

The input files produced by Chrome seem to be sliced in arbitrary locations - they don't end cleanly at the boundaries of EBML tags.  I did find a bug in the library that made this situation harder to decipher (pushed up a fix in version 0.4.1), but that by itself doesn't help your situation  ...well, it might a little since there isn't a panic anymore.  **However**, if you concatenate the files produced by Chrome together, they _can_ be read cleanly (well, up until the end of the last file.  Unless there are additional segments that aren't in the example?). 

The files from Firefox seem to be sliced at tag boundaries, which is why you weren't having any trouble parsing them.  (Although it is a little strange to see webm files that start out with SimpleBlocks outside of any Cluster tag).

tldr; I think if you chain/concat your inputs together you should be golden 👍

**Comment by Innominus** [comment]
Really appreciate you digging! I did not know that, because I guess when it's not a contiguous slice the webm-iterable parser tags would just give me rawtags and no cluster tags or anything like that..

Unfortunately the idea of this webm restorer is to take the header tags from the first file and plop them on top of cluster tags. 
Cluster tags supposedly with the MediaRecorder spec will always have a keyframe, which makes life a lot easier.

When you're talking about Chrome slicing in arbitrary locations, you don't mean a clean segment tag or cluster tag end, you mean the tag itself won't be readable with a single file as it's cut in half or so and put into the next file?
I haven't tried with 0.4.1, but all I really need is to be able to read a cluster tag start and then search through the next files coming through for a cluster end. Is my best bet getting the first two files and stitching them together, then waiting for the next two files to come through and work in blocks of two files, instead of the one file I was doing with FF? 

Also, how come the tags were showing as RawTags if they're actually SimpleBlocks? Is this because they weren't between cluster or segments or?

If I can now read the tags coherently and not get a panic with the fix, I should be able to remedy this myself! I will have a go. Appreciate any input you have on my process though with Chrome. Thanks :)

edit: Just to add, yes there is an additional segment missing, and that segment is when the mediarecorder stops playing, I'm not capturing that final segment data.

**Comment by Innominus** [comment]
Just to add, I'm getting weird memory leak stuff. 
I can't recreate what you've done and read the Chrome output. I still get errors, and when I try to skip over the errors, I get memory core dumps.
This is the function that runs over the combined Chrome output:
`pub fn retrieve_header_tags(header_file: File) -> Result<Vec<MatroskaSpec>, TagIteratorError> {
    let mut tags = Vec::new();

    for res in WebmIterator::new(header_file, &[]) {
        let tag = match res {
            Ok(matroska) => matroska,
            Err(e) => {
                // println!("{e}");
                continue;
            }
        };

        if matches!(
            tag,
            MatroskaSpec::SimpleBlock(_)
                | MatroskaSpec::Timestamp(_)
                | MatroskaSpec::Cluster(_)
                | MatroskaSpec::Segment(Master::End)
        ) {
            continue;
        }

        println!("{tag:?}");
        tags.push(tag);
    }

    Ok(tags)
}`

And this is the error I'm getting from output:
`MaxBlockAdditionID(1)
CodecID("V_VP9")
Video(Start)
PixelWidth(640)
PixelHeight(480)
AlphaMode(1)
Video(End)
TrackEntry(End)
Tracks(End)
memory allocation of 562945658388478 bytes failed
Aborted (core dumped)
`
That's like 562k gigabytes aha not sure what's happening here. It's hitting the error multiple times maybe? Can't iterate over it?

**Comment by austinleroy** [comment]
Hey there! I'll try and answer your questions as best I can:

--------------------
> When you're talking about Chrome slicing in arbitrary locations, you don't mean a clean segment tag or cluster tag end, you mean the tag itself won't be readable with a single file as it's cut in half or so and put into the next file?
I haven't tried with 0.4.1, but all I really need is to be able to read a cluster tag start and then search through the next files coming through for a cluster end. Is my best bet getting the first two files and stitching them together, then waiting for the next two files to come through and work in blocks of two files, instead of the one file I was doing with FF?

So EBML (which is the encoding format for webm files) doesn't have typical start/end tags encoded into it like html or xml.  The format itself is very simplistic - at a high level, the whole format can be expressed as `<Tag Id><Tag Size><Tag Data>`.  `Tag Id` and `Tag Size` are encoded an vInts (variable integer; 1-8 bytes read as a number slightly differently than a regular integer), and `Tag Data` is raw binary.  Given this format, unfortunately there isn't a sane way to start reading a file from an arbitrary position (since there are no markers or escape sequences to identify the start of a tag).  The best that could be done is to move through a file byte-by-byte until a valid `Tag Id` is found, but there's no way to ensure correctness, as that could easily be a part of regular binary data inside a `Tag Data`.  Without having a known starting point and known length, there isn't a way to ensure tag parsing will occur at the correct place in the file.

--------------------
> Also, how come the tags were showing as RawTags if they're actually SimpleBlocks? Is this because they weren't between cluster or segments or?

You were seeing `RawTag` variants because the leading vInt in files after the first didn't match any known Tag Ids in the matroska spec.

--------------------

For your most recent comment about the memory leaks - the success I had was by combining each sequential file in order, i.e. `cat wire-1_segment_0.webm wire-1_segment_1.webm wire-1_segment_2.webm > combined.webm`.  Without starting at the beginning of a valid tag, you're likely to run into weird errors similar to the one you posted (reading a tag size of 562Tb)

I'll try and continue to think about the best way to solve this... sorry you haven't had any luck yet!

**Comment by austinleroy** [comment]
I just did a little more digging on the MediaRecorder API and found the following in [the w3c spec](https://w3c.github.io/mediacapture-record/#mediarecorder-methods):

> The UA MUST record [stream](https://w3c.github.io/mediacapture-record/#dom-mediarecorder-stream) in such a way that the original Tracks can be retrieved at playback time. When multiple [Blob](https://w3c.github.io/FileAPI/#dfn-Blob)s are returned (because of [timeslice](https://w3c.github.io/mediacapture-record/#dom-mediarecorder-start-timeslice-timeslice) or [requestData()](https://w3c.github.io/mediacapture-record/#dom-mediarecorder-requestdata)), the individual Blobs need not be playable, but the combination of all the Blobs from a completed recording MUST be playable.

Given this, it sounds like Chrome isn't necessarily doing anything wrong by providing file parts sliced outside natural tag boundaries - the spec only insists that the combination of all Blobs from a recording should work.

I'm thinking about adding a new TagIteratorError variant specifically for `UnexpectedEOF` when reading a tag.  I'm thinking it will contain partial tag data of what could be read along with the start index/byte offset of the tag that ended unexpectedly.  For your use case, you could use this data to combine the files intelligently rather than having to combine the entire recording into a single file before trying to parse.

If I make such an update, some psuedo-code for your use case (it'll definitely need some clean up) would then be:

```
pub fn build_files() {
    let mut Vec<MatroskaSpec> header_tags = Vec::new();
    let mut Vec<u8> previous_file_remainder = Vec::new();
    for file in filesOrderedByCreation {
        let file_tags = process_file(file, &mut previous_file_remainder, &mut header_tags)?;
        // write file_tags to a new file or something
    }
}

pub fn process_file(file: File, previous_file_remainder: &mut Vec<u8>, header_tags: &mut Vec<MatroskaSpec>) -> Result<Vec, TagIteratorError> {
    let mut file_tags: Vec<MatroskaSpec> = Vec::new();
    let file_segment = previous_file_remainder.clone().chain(file);
    for res in WebmIterator::new(file_segment, &[]) {
        let tag = match res {
            Ok(matroska) => matroska,
            Err(e) => {
                if let TagIteratorError::UnexpectedEOF(pos) = e {
                    file.seek(SeekFrom::Start(pos));
                    file.read_to_end(&mut previous_file_remainder);
                    break;
                }
                return Err(e);
            }
        };

        if matches!(
            tag,
            MatroskaSpec::SimpleBlock(_)
                | MatroskaSpec::Timestamp(_)
                | MatroskaSpec::Cluster(_)
                | MatroskaSpec::Segment(Master::End)
        ) {
            file_tags.push(tag);
            continue;
        } else {
            println!("{tag:?}");
            header_tags.push(tag);
        }
    }
    let mut result = header_tags.clone().
    result.append(file_tags);
    Ok(result)
}
```

What do you think of this approach?  If you like it, I can start working on updating ebml-iterable and can probably finish it within a few days.

**Comment by Innominus** [comment]
Hey! Thanks so much for getting back to me once again. That information is really valuable and helpful.
That approach you've put to handling unexpected EOF is something I was theory crafting doing but didn't know how to begin working on. Working on files of 2 or more, and holding up the engine until I can safely complete a cluster and create a readable file would be awesome! I would very much appreciate if you could add that to the API :) thanks once again for getting back to me.

**Comment by austinleroy** [comment]
I was able to put together a change that is close to what I suggested above (I added a couple more properties to the error that may or may not be helpful).  Hopefully this will get you where you need to be!  Let me know if you run into any more issues and I can try and take a stab at them as time permits :)

Edit:  Forgot to mention that this change is in version 0.4.2

**Comment by austinleroy** [comment]
I'm going to assume these changes worked for you - if not, feel free to reopen this issue or open a new one.  Thanks again for your interest!

**Comment by friofry** [comment]
Hi Austin, 

I just tried to use webm-iterator for webm created by MediaRecorder:
Here is the function that tries to set Duration if it is not set.

```rust
pub fn set_webm_duration_if_not_exist(file_path: &str, dst_file_path: &str, duration_ms: f64, timestamp_scale_opt: Option<u64>) -> Result<(), Box<dyn std::error::Error>> {
    let timestamp_scale = timestamp_scale_opt.unwrap_or(1000000);
    let mut src = File::open(file_path)?;
    let mut tag_iterator = WebmIterator::new(&mut src, &[MatroskaSpec::Info(Master::Start)]);

    let mut dest = File::create(dst_file_path)?;
    let mut tag_writer = WebmWriter::new(&mut dest);

    for tag in tag_iterator {
        match tag {
            Ok(tag) => {
                match tag {
                    MatroskaSpec::Info(info) => {
                        let children = info.get_children();

                        let duration_tag = children.iter().find(|&child| matches!(child, MatroskaSpec::Duration(_)));
                        let timestamp_scale_tag = children.iter().find(|&child| matches!(child, MatroskaSpec::TimestampScale(_)));

                        let mut children = children.clone();
                        if duration_tag.is_none() {
                            children.push(MatroskaSpec::Duration(duration_ms));
                        }
                        if timestamp_scale_tag.is_none() {
                            children.push(MatroskaSpec::TimestampScale(timestamp_scale));
                        }
                        tag_writer.write(&MatroskaSpec::Info(Master::Full(children)))?;
                    }
                    _ => {
                        tag_writer.write(&tag)?;
                    }
                }
            },
            Err(e) => {
                println!("error {:?}", e);
                tag_writer.flush()?;
                break;
            }
        }
    }
    println!("flush");
    tag_writer.flush()?;
    Ok(())
}
```

The last tag in the test webm file is corrupted. Here is the output of mkvdump:
```
...
    - id: SimpleBlock
      header_size: 2
      size: 14
      value:
        track_number: 1
        timestamp: 2340
        keyframe: true
    - id: SimpleBlock
      header_size: 2
      size: 118
      value:
        track_number: 2
        timestamp: 2365
    - id: Corrupted
      header_size: 0
      size: 1
      value: null
```

`tools::read_vint` returns a `size` equal to `4299262263296` in the `peek_valid_tag_header` method.
The application then panics `EBMLSize::new` with that size.

I suppose in this case the webm iterator should throw an `UnexpectedEOF` error

**Comment by austinleroy** [comment]
Hey there Andrey,

Would you be able to provide the test webm or a link to it?  I'd like to try debugging it locally.  Looking through the code I'm not seeing any obvious spots that would panic inside EBMLSize::new().

**Comment by friofry** [comment]
Hi @austinleroy can I send it to you via DM or email?

**Comment by austinleroy** [comment]
Sure, feel free to send an email with it (you can find my email in my profile)

**Comment by austinleroy** [comment]
So the error I'm seeing pops up when trying to read tag data for the tag with size `4299262263296`.  The tag is a SimpleBlock (binary) tag, so there isn't currently a limit on the expected size.  When the code tries to allocate a Vec to hold that tag, it obviously panics since I don't have 4.3Tb of memory.  While it's easy to say "4.3Tb is clearly an invalid size", it's harder to define exactly where the threshold should be before deciding a tag size is "corrupted".  I went ahead and decided on a default limit of 4GB, which can be overridden as necessary on the `WebmIterator`.  This will now throw an `CorruptedFileData::InvalidTagSize` error, which your code will catch and print.  You can pull this update in version 0.6.1.  Feel free to reach out if you have any more thoughts on this or run into any other issues!

**Comment by friofry** [comment]
Hi @austinleroy 
Thanks for adding CorruptedFileData::InvalidTagSize, that helps.

But now I have another panic:

```
thread '<unnamed>' panicked at ...\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ebml-iterable-0.6.1\src\tag_iterator.rs:240:29:
slice index starts at 3 but ends at 1
```
(I've emailed you the video file)

**Comment by austinleroy** [comment]
Hey @friofry 
This panic should be fixed in version 0.6.2.  That was a really weird combination of specific details in the test file that caused that one (basically the EOF lined up with buffered memory in such a way that the existing EOF checks didn't trigger).  Sorry it took a little longer to get this one addressed

**Comment by friofry** [comment]
Thanks! updating to 0.6.2
